File: source/bindings/java/com/microsoft/cognitiveservices/speech/translation/TranslationRecognitionResult.java
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//
package com.microsoft.cognitiveservices.speech.translation;

import java.util.HashMap;
import java.util.Map;

import com.microsoft.cognitiveservices.speech.RecognitionResult;
import com.microsoft.cognitiveservices.speech.util.Contracts;
import com.microsoft.cognitiveservices.speech.util.SafeHandle;
import com.microsoft.cognitiveservices.speech.util.SafeHandleType;
import com.microsoft.cognitiveservices.speech.util.StringMapRef;

/**
  * Defines translation recognition result.
  */
//FIXME is this ok to make not final?
public class TranslationRecognitionResult extends RecognitionResult {

    //FIXME is this ok to make public?
    /**
     * Constructor
     * @param result
     */
    public TranslationRecognitionResult(long result) {
        super(result);
        Contracts.throwIfNull(resultHandle, "resultHandle");
        StringMapRef translationsRef = new StringMapRef();
        Contracts.throwIfFail(getTranslations(resultHandle, translationsRef));
        this.translations = new HashMap<String, String>(); 
        if (!translationsRef.getValue().isEmpty()) {
            this.translations.putAll(translationsRef.getValue());
        }
    }

    /**
     * Presents the translation results. Each item in the dictionary represents translation result in one of target languages, where the key
     * is the name of the target language, in BCP-47 format, and the value is the translation text in the specified language.
     * @return the current translation map.
     */
    public final Map<String, String> getTranslations() {
        return this.translations;
    }

    /**
     * Returns a String that represents the translation recognition result.
     * @return A String that represents the translation recognition result.
     */
    @Override
    public String toString() {
        String text = "ResultId:" + this.getResultId() +
                " Reason:" + this.getReason()  +
                ", Recognized text:<" + this.getText() + ">.\n";

        for(String key : this.translations.keySet()) {
            text += "    Translation in " + key + ": <" + this.translations.get(key) + ">.\n";
        }

        return text;
    }

    private final native long getTranslations(SafeHandle resultHandle, StringMapRef translations);

    private Map<String, String> translations;
}

File: source/bindings/csharp/translation_recognition_result.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Runtime.InteropServices;
using Microsoft.CognitiveServices.Speech.Internal;
using static Microsoft.CognitiveServices.Speech.Internal.SpxExceptionThrower;

namespace Microsoft.CognitiveServices.Speech.Translation
{
    /// <summary>
    /// Defines a translation result.
    /// </summary>
    public class TranslationRecognitionResult : RecognitionResult
    {
        internal TranslationRecognitionResult(IntPtr resultPtr) : base(resultPtr)
        {
            GetTranslationTexts();
        }

        /// <summary>
        /// Presents the translation results. Each item in the dictionary represents translation result in one of target languages, where the key
        /// is the name of the target language, in BCP-47 format, and the value is the translation text in the specified language.
        /// </summary>
        public IReadOnlyDictionary<string, string> Translations { get { return translationTextResultMap; } }

        /// <summary>
        /// Returns a string that represents the speech recognition result.
        /// </summary>
        /// <returns>A string that represents the speech recognition result.</returns>
        public override string ToString()
        {
            var text = string.Format(CultureInfo.InvariantCulture, "ResultId:{0} Reason:{1}, Recognized text:<{2}>.\n", ResultId, Reason, Text);
            foreach (var element in Translations)
            {
                text += $"    Translation in {element.Key}: <{element.Value}>.\n";
            }
            return text;
        }

        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Globalization", "CA1303", Justification = "exceptions not localized")]
        private void GetTranslationTexts()
        {
            ThrowIfNull(resultHandle, "Invalid result handle");

            if (Internal.RecognitionResult.recognizer_result_handle_is_valid(resultHandle))
            {
                int bufLen = 0;
                translationTextResultMap = new Dictionary<string, string>();
                IntPtr hr = Internal.RecognitionResult.translation_text_result_get_translation_text_buffer_header(resultHandle, IntPtr.Zero, ref bufLen);
                if (hr == SpxError.BufferTooSmall)
                {
                    IntPtr translationResultBuffer = Marshal.AllocHGlobal(bufLen);
                    try
                    {
                        ThrowIfFail(Internal.RecognitionResult.translation_text_result_get_translation_text_buffer_header(resultHandle, translationResultBuffer, ref bufLen));

                        int offset = 0;
                        int bufferSize = (int)Marshal.ReadIntPtr(translationResultBuffer, offset);

                        ThrowIfFail(bufferSize == bufLen, SpxError.RuntimeError.ToInt32());

                        offset += IntPtr.Size;
                        int numberEntries = (int)Marshal.ReadIntPtr(translationResultBuffer, offset);

                        offset += IntPtr.Size;
                        IntPtr targetLanguagesPtr = Marshal.ReadIntPtr(translationResultBuffer, offset);

                        offset += IntPtr.Size;
                        IntPtr translationTextsPtr = Marshal.ReadIntPtr(translationResultBuffer, offset);

                        // read targetLanguages and translationTexts
                        for (int entryIndex = 0; entryIndex < numberEntries; entryIndex++)
                        {
                            offset = IntPtr.Size * entryIndex;
                            IntPtr targetLanguageStrPtr = Marshal.ReadIntPtr(targetLanguagesPtr, offset);
                            string language = Utf8StringMarshaler.MarshalNativeToManaged(targetLanguageStrPtr);
                            IntPtr translationTextStrPtr = Marshal.ReadIntPtr(translationTextsPtr, offset);
                            string text = Utf8StringMarshaler.MarshalNativeToManaged(translationTextStrPtr);
                            translationTextResultMap[language] = text;
                        }
                    }
                    finally
                    {
                        Marshal.FreeHGlobal(translationResultBuffer);
                    }
                }
            }
        }

        private Dictionary<string, string> translationTextResultMap;
    }
}
